lib/src/over_react_test/custom_matchers.dart

```diff
@@ -337,3 +339,261 @@ Matcher throwsPropError_Combination(String propName, String prop2Name, [String m
       ))
   ));
 }
+
+/// A log matcher that ensures that expected log values are present.
+///
+/// In its core, the matcher compares two lists. The added utility is that by
+/// passing in a callback, the matcher will run the function and record the logs,
+/// return them in a list, and use that list as the actual value to compare
+/// against the expected.
+///
+/// Related: [recordConsoleLogs]
+class LogMatcher extends Matcher {
```

Comment:
An idea that occurred to me while reviewing your PR and the complexity of the matcher implementation. I think you might also be able to composing matchers and leverage existing iterable matchers to accomplish similar functionality; have you experimented at all with this approach?

This could cut down on complexity and code, and add in some flexibility to handle other cases.

Something like:
```dart
class LoggingFunctionMatcher extends CustomMatcher {
  final Matcher _matcher;

  /// [matcher] is tested against the list of recorded logs
  // could make wrapMatcher be `contains` by default as well if desired
  LoggingFunctionMatcher(dynamic matcher) : super(..., wrapMatcher(matcher))

  @override
  featureValueOf(actual) {
    if (actual is! Function()) {
      throw ArgumentError();
    }
  
    startRecordingLogs();
    try {
      actual();
    } catch (_) {}
    stopRecordingLogs();

    return recordedLogs;
  }
}
```

```dart
Matcher emitsLogs(dynamic expected) => LoggingFunctionMatcher(expected);
// could even make this inclusive by default:
// Matcher emitsLogs(dynamic expected) => LoggingFunctionMatcher(expected is Matcher ? expected : containsAll(expected));
Matcher emitsSingleLog(dynamic expected) => LoggingFunctionMatcher([expected]);
Matcher emitsLog(dynamic expected) => LoggingFunctionMatcher(contains(expected));

final Matcher emitsNoLogs => LoggingFunctionMatcher(isEmpty);
```
```dart
expect(() => ..., emitsLog('This is the whole damn log');
expect(() => ..., emitsLog(contains('substring of the log'));
expect(() => ..., emitsNoLogs);
expect(() => ..., emitsLogs([
  contains('prop FooProps.hey is required'),
  contains('prop FooProps.what is required'),
]);
// Can even use other iterable matchers
expect(() => ..., emitsLogs(containsAll([
  contains('prop FooProps.hey is required'),
  contains('prop FooProps.what is required'),
]));
expect(() => ..., emitsLogs(hasLength(5));
```